### ISSUE TITLE
Trying to use cuSOLVERMp on Perlmutter

### ISSUE THREAD

**Issue by s769** [NVIDIA/CUDALibrarySamples#226]
I am trying to use the cuSOLVERMp library on Perlmutter. I built the examples and tried to run them on a GPU node with

`srun -n 2 ./mp_potrf_potrs -verbose 1`

The output is attached in out.txt. I also set the environment variables as mentioned in issue #190 
(this issue is from when I was trying to use the library on another system). I also tried setting `OMPI_MCA_coll_ucc_enable=0` based on the comments in HPC-SDK [docs](https://docs.nvidia.com/hpc-sdk//hpc-sdk-release-notes/index.html)

I am finding that the program runs successfully sometimes, but errors other times. The main error I get is

```
ucc_coll.c:239 UCC ERROR failed to init collective: Resources are not available for the operation
mp_potrf_potrs: mp_potrf_potrs.c:488: main: Assertion `cusolverStat == CUSOLVER_STATUS_SUCCESS' failed.

```

Here is my module list:

```
1) craype-x86-milan 8) gpu/1.0 15) evp-patch
2) libfabric/1.15.2.0 9) craype/2.7.30 (c) 16) python/3.11 (dev)
3) craype-network-ofi 10) cray-dsmml/0.2.2 17) cudatoolkit/12.4 (g)
4) xpmem/2.6.2-2.5_2.38__gd067c3f.shasta 11) cray-libsci/23.12.5 (math) 18) nvidia/24.5 (g,c)
5) perftools-base/23.12.0 12) PrgEnv-nvidia/8.5.0 (cpe) 19) cray-mpich/8.1.28 (mpi)
6) cpe/23.12 13) cray-hdf5-parallel/1.12.2.3 (io)
7) craype-accel-nvidia80 14) conda/Miniconda3-py311_23.11.0-2

```

I have opened up a ticket at NERSC but wanted to check here as well in case people have seen similar things before.

[out.txt](https://github.com/user-attachments/files/17382781/out.txt)

**Comment by s769** [comment]
From my tests, I found that when running the example with options `-p 1 -q <num_procs`, it completes successfully. However, using a value of `-p` greater than 1 leads to the error (sometimes it still works, but mostly it fails). 

I am also finding that when running on multiple nodes (e.g. `srun -n 8 ./mp_potrf_potrs -p 1 -q 8 ...`), the program seems to hang. I have made sure that num_gpus = p * q. I also tried setting/unsetting the OMPI_MCA_coll_ucc_enable=0 variable, but this does not affect it.

**Comment by mrogowski** [comment]
Can you check if setting the `UCC_TL_NCCL_LAZY_INIT` environment variable to `no` makes any difference?

**Comment by s769** [comment]
That seemed to do the trick, thanks! (That was the only env var I needed to set for the example to run correctly)

Is there anywhere I can get a list of these environment variables and what they do? Would be nice to have in case I need to build on another system again.

**Comment by mrogowski** [comment]
@s769, we do not have a comprehensive list. cuSOLVERMp depends on UCC, NCCL, etc., and issues/incompatibilities occur once in a while in some of the dependencies. Having said that, we will be adding the "Troubleshooting and Known issues" section to the documentation soon to list known issues and workarounds such as the `UCC_TL_NCCL_LAZY_INIT` variable.

**Comment by s769** [comment]
Reopening this because I noticed that sometimes, the `mp_potrf_potrs` example will error at the assertion of `cusolverStat == CUSOLVER_STATUS_SUCCESS` after the `potrf` call. This appears to happen randomly, and I'm not sure what could be causing it. It's saying that the `cusolverStat` is 7, which I believe corresponds to `CUSOLVER_STATUS_MATRIX_TYPE_NOT_SUPPORTED`, but I don't know how that could be getting affected here.

**Comment by s769** [comment]
I believe I found the reason for this problem. I think that the matrix generation function sometimes generates a matrix that is close to being semidefinite, which causes the factorization to error. This also explains why it was happening randomly, since the matrix elements are randomly generated. I think I have fixed this by just setting the diagonal elements to `n` (or even `10*n` to be extra sure) instead of `2*sum` (since each element will be in [0,1), the sum has to be less than `n`).

**Comment by mrogowski** [comment]
Thanks for following up and answering your own question! 🙂
I think you are right and the matrix generator we used here can generate a matrix that is not SPD. We are tracking this internally and will fix it in the future.
To verify, you can see if you call `srand(x)` with a fixed value of `x` you can get it to pass/fail reliably.

**Comment by s769** [comment]
I saw that it's easier to get it to fail for larger matrix sizes. Here's an example where I got it to fail:

I inserted the line `srand(opts.m)` before the `generate_diagonal_dominant_symmetric_matrix(n, _A, lda)` line. I used `opts.m` as the seed since it was an easy way to change the seed from the command line, and I believe `m` is not used otherwise in the `mp_potrf_potrs` program.

Then, running `mpirun -n 4 ./mp_potrf_potrs -p 4 -q 1 -ia 1 -ja 1 -ib 1 -jb 1 -mbA 500 -mbB 500 -nbA 500 -nbB 500 -n 1000 -m 10` led to the error after the `potrf` call.
